File: tree/543.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
    if root == nil {
        return 0
    }
    var ret int
    dfs(root, &ret)
    return ret - 1
}

func dfs(root *TreeNode, ret *int) int {
    if root == nil {
        return 0
    }
    left := dfs(root.Left, ret)
    right := dfs(root.Right, ret)
    if left + right + 1 > *ret {
        *ret = left + right + 1
    }
    if left > right {
        return left + 1
    }else {
        return right + 1
    }
}

File: tree/543.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if (!root)
            return 0;
        ret = std::max(ret, recur(root));
        return ret-1;
    }
private:
    int recur(TreeNode* root) {
        if (!root)
            return 0;
        int left = recur(root->left);
        int right = recur(root->right);
        ret = std::max(ret, left + right + 1);
        return std::max(left, right) + 1;
    }
    int ret = 0;
};
